README.md
JJ's DH VPS Memory Manager
=========================

**IMPORTANT NOTE: While I work for DreamHost this application is NOT supported by DreamHost and any questions should be coming to me NOT DH Support!**

This application was built to be a VPS Memory Manager focused on uptime rather then cost savings. You can see one of my initial post on the subject here: http://www.gimmesoda.com/thoughts-on-vps-memory-management/

Installation
-----------

1. Grab the files using git or download them from github: https://github.com/jgalvez/JJs-DH-VPS-Memory-Manager
2. Upload the files.
3. Create an API key for use with the memory manager from here: https://panel.dreamhost.com/index.cgi?tree=home.api&

The key must have access to the following two functions:

dreamhost_ps-set_size
services-progress

4. Visit the URL you uploaded the memory manager, the install script will start.
5. Fill out the form that comes up, save the settings.

Configuration
------------

Once installed you should fine tune the application configuration. Let's look at the important values in the config.php file:

define('MIN_MEMORY', '300');

define('MAX_MEMORY', '4000');

MIN_MEMORY should be the minimum memory your VPS uses. 300 is default only because that is the lowest amount of memory you are allowed to use, change this to something more realistic.

MAX_MEMORY should be a maximum amount of memory you are willing to pay for, remember that the memory manager will only resize to this level if the memory is actually needed. You are not likely to be at this level of memory usage for long periods of time, if you are go buy a dedicated instead.

SAFETY_PERCENT can normally be left at the default value, however if you want to get experimental this value is basically the amount of memory (as a percentage of total memory) which the application will attempt to keep available.

EMAIL_ON_RESIZE: Do you want to be emailed on resize? Set to true if you do!

TWEET_ON_RESIZE: Do you want the application to send a tweet on resize? Set to true and fill out the following variables if you do!

TWEET_CONSUMER_KEY
TWEET_CONSUMER_SECRET
TWEET_OAUTH_TOKEN
TWEET_OAUTH_SECRET

You will need to register with Twitter to get the needed values to these variables: https://dev.twitter.com/apps/new

When you register make sure that your selected access level's are set to "Read and write".

Upgrading!
-----------------
Until I work out a decent auto-upgrade feature there are currently two ways to get upgrades in place.

1) Download the new files and upload them over the current files (no logs should be lost and your config will remain in place)
2) If you used 'git clone' to grab the application just run 'git pull' from the application's directory.

Next you'll want to make a copy of your config.php file, then delete the copy on your server. Visit the memory manager url and you'll see the installation page, copy your API key from the old config and enter the user/pass and email address you want to use. The new config with the new definitions will be written by the application.

Once that is done restart the application daemon as explained below.

Restart the Daemon
-----------------

Anytime you make a change to the configuration file you should restart the daemon (using the restart link within the application).

That's it! Feel free to contact me at jj@gimmesoda.com with any questions.

Credits and Thanks
-----------------

Many thanks to the various people who have contributed their time and efforts in getting this application tested, released, and themed. In no particular order:

DreamHost: For giving me such an amazing career that I want to work on things like this on my free time.

GY'ers: An extended family, thanks for putting up with all the rants and ramblings about VPS services and everything else!

Tyler Brekke: Theme development/implementation, and an awesome co-worker.

Kan Adachi: Thanks for putting your amazing skills to work and making this such a polished looking product.

Samuel Corral: Testing and encouragement from my earliest versions, another amazing co-worker.

Sarah Galvez: Even after all these years, you still put up (and love me) while so much of my free time is spent on development. <3

CHANGELOG.md
## 1.2.0 (2013-03-15)
* refactor: major code cleanup of the daemon.php, new MemoryManagerDaemon class
* bugfix: do not attempt resize below current actual memory usage
* bugfix: we should no longer resize down until at least 30 minutes has gone by from last resize 
* bugfix: cronjobs added properly for those who never entered a cronjob before (adds MAILTO="" line)
* feature: new configurable for seconds before decrease, use SECONDS_BEFORE_DECREASE in config.php defaults to 1800
* feature: new configurable for minutes between log entries used in graphing data, use LOG_EVERY_X_MINUTES in the config.php defaults to 5
* feature: new configurable for seconds between memory checks, use SECONDS_TO_WAIT_BETWEEN_MEMORY_CHECKS in the config.php defaults to 5
* feature: new configurable for number of checks on API progress before giving up, use MAX_CHECKS_BEFORE_GIVING_UP in the config.php defaults to 50
* feature: new configurable for seconds between API progress checks, use SECONDS_TO_WAIT_BETWEEN_RESIZE_STATUS_CHECKS in the config.php defaults to 5
* feature: new configurable for the daemon's timezone, use DAEMON_TIMEZONE in the config.php defaults to 'America/Los_Angeles'. Check your server's /usr/share/zoneinfo for supported timezones.
* extra: added a .gitignore for cache, run files, logs, and config.php

## 1.1.3 (2012-07-24)
* bugfix: uid/gid was being set incorrectly in some rare cases
* extra: formatting fixes

## 1.1.2 (2012-04-27)
* extra: removed work around for the now valid api.dreamhost.com cert
* extra: readme updated to explain SAFETY_PERCENTAGE

## 1.1.1 (2012-04-14)
* bugfix: daemon now cleans log
* bugfix: properly loading excanvas for IE
* extra: consolidated jquery/flot files
* extra: updated flot library	
* workaround: ignoring expired api.dreamhost.com cert

## 1.1.0 (2011-12-09)
* feature: process snapshot when change is requested
* feature: tweet on resize (uses oauth: see readme)
* feature: email on resize
* bugfix: cached memory was being reported in kb not mb
* bugfix: crontab needs a newline to properly RELOAD
* bugfix: prevent large error log when using cron
* bugfix: was looping through missing array when login.php does not exist
* extra: Added CHANGELOG

## 1.0.0 (2011-12-01)

* First stable release

config_sample.php
<?php

if(basename($_SERVER['SCRIPT_FILENAME']) == 'config.php') {
  die('You may not access the config file directly.');
}

# Required Settings
define('DH_API_KEY', '--DHAPIKEY--');
define('USER', '--USER--');
define('PASS', '--PASS--');
define('SALT', '--SALT--');
define('EMAIL', '--EMAIL--');

# Application Tuning
define('HOSTNAME', '--HOSTNAME--');
define('MIN_MEMORY', '--MIN_MEMORY--');
define('MAX_MEMORY', '--MAX_MEMORY--');
define('SAFETY_PERCENT', '--SAFETY_PERCENT--');
define('TEMPLATE', '--TEMPLATE--');
define('ALWAYS_USE_COMMITTED_AS', '--ALWAYS_USE_COMMITTED_AS--');
define('LOG_ALL', '--LOG_ALL--');
define('IGNORE_CACHE', '--IGNORE_CACHE--');

# Daemon Settings
define('DAEMON_USER', '--DAEMON_USER--');
define('DAEMON_GROUP', '--DAEMON_GROUP--');
define('CHANGE_MEMORY', '--CHANGE_MEMORY--');
define('SECONDS_BEFORE_DECREASE', 1800);
define('LOG_EVERY_X_MINUTES', 5);
define('MAX_CHECKS_BEFORE_GIVING_UP', 50);
define('SECONDS_TO_WAIT_BETWEEN_MEMORY_CHECKS', 5);
define('SECONDS_TO_WAIT_BETWEEN_RESIZE_STATUS_CHECKS', 5);
define('DAEMON_TIMEZONE', 'America/Los_Angeles');

# Contact Settings
define('EMAIL_ON_RESIZE', '--EMAIL_ON_RESIZE--');
define('TWEET_ON_RESIZE', '--TWEET_ON_RESIZE--');

# Tweet Settings
# Register the Memory Manager with Twitter
# https://dev.twitter.com/apps/new
define('TWEET_CONSUMER_KEY', '--TWEET_CONSUMER_KEY--');
define('TWEET_CONSUMER_SECRET', '--TWEET_CONSUMER_SECRET--');
define('TWEET_OAUTH_TOKEN', '--TWEET_OAUTH_TOKEN--');
define('TWEET_OAUTH_SECRET', '--TWEET_OAUTH_SECRET--');

?>